bouncer/shared/fill_or_kill.ts

```diff
@@ -0,0 +1,139 @@
+    asset,
+    swapAsset,
+    destAddress,
+    'FoK_Test',
+    undefined, // messageMetadata
+    0, // brokerCommissionBps
+    false, // log
+    0, // boostFeeBps
+    refundParameters,
+  );
+  const depositAddress = swapRequest.depositAddress;
+  const depositChannelId = swapRequest.channelId;
+
+  const observeDestinationAddress =
+    asset === Assets.Dot ? decodeDotAddressForContract(destAddress) : destAddress;
+
+  const observeSwapScheduled = observeEvent('swapping:SwapScheduled', {
+    test: (event) =>
+      typeof event.data.origin === 'object' &&
+      'DepositChannel' in event.data.origin &&
+      event.data.origin.DepositChannel.channelId === depositChannelId.toString() &&
+      event.data.destinationAddress[shortChainFromAsset(swapAsset)]?.toLowerCase() ===
+        observeDestinationAddress.toLowerCase(),
+  }).event;
+
+  // Deposit the asset
+  await send(asset, depositAddress, amount.toString());
+  console.log(`Sent ${amount} ${asset} to ${depositAddress}`);
+
+  const swapScheduledEvent = await observeSwapScheduled;
+  const swapId = Number(swapScheduledEvent.data.swapId.replaceAll(',', ''));
+  console.log(`${asset} swap scheduled, swapId: ${swapId}`);
+
+  // TODO: Observing after the SwapScheduled event means its possible to miss the events, but we need to the swap id.
+  const observeRefundEgressScheduled = observeEvent(`swapping:RefundEgressScheduled`, {
+    test: (event) => Number(event.data.swapId.replaceAll(',', '')) === swapId,
+  }).event;
+  const observeSwapExecuted = observeEvent(`swapping:SwapExecuted`, {
+    test: (event) => Number(event.data.swapId.replaceAll(',', '')) === swapId,
+  }).event;
+  const observeSwapRescheduled = observeEvent(`swapping:SwapRescheduled`, {
+    test: (event) => Number(event.data.swapId.replaceAll(',', '')) === swapId,
+  }).event;
+
+  const firstSwapEvent = await Promise.race([
+    observeSwapRescheduled,
+    observeSwapExecuted,
+    observeRefundEgressScheduled,
+  ]);
+  if (firstSwapEvent.name.method === 'SwapExecuted') {
+    throw new Error(`${asset} swap ${swapId} was executed instead of failing and being retried`);
+  } else if (firstSwapEvent.name.method !== 'SwapRescheduled') {
+    throw new Error(`${asset} swap ${swapId} was not retried once as expected`);
+  }
+  console.log(`${asset} swap ${swapId} was rescheduled`);
+
+  const secondSwapEvent = await Promise.race([observeRefundEgressScheduled, observeSwapExecuted]);
+  if (secondSwapEvent.name.method === 'SwapExecuted') {
+    throw new Error(`${asset} swap ${swapId} was executed instead of being refunded`);
+  }
+
+  const refundBalanceAfter = await observeBalanceIncrease(
+    asset,
+    refundAddress,
+    refundBalanceBefore,
+  );
+
+  console.log(
+    `Refund balance before: ${refundBalanceBefore}, after: ${refundBalanceAfter} ${asset} `,
+  );
+  // We expect the refund to be a little less due to ingress and egress fees.
+  assert(
+    isWithinOnePercent(
```

Comment:
I think removing this 1% margin is going to reduce the chances of future flakiness a lot

Reply:
I assume you mean that just observing a balance increase at all is fine for this test.

Reply:
yeah exactly, especially if we use a random address, then it will be fine